tests/cuda/test_barrier.cu

```diff
@@ -0,0 +1,158 @@
+ */
+
+#include <gtest/gtest.h>
+
+#include "../../device/cuda/src/utils/barrier.hpp"
+
+__global__ void testBarrierAnd(bool* out) {
+    traccc::cuda::barrier bar;
+
+    bool v;
+
+    v = bar.blockAnd(false);
+    if (threadIdx.x == 0) {
+        out[0] = v;
+    }
+
+    v = bar.blockAnd(true);
+    if (threadIdx.x == 0) {
+        out[1] = v;
+    }
+
+    v = bar.blockAnd(threadIdx.x % 2 == 0);
+    if (threadIdx.x == 0) {
+        out[2] = v;
+    }
+
+    v = bar.blockAnd(threadIdx.x < 32);
+    if (threadIdx.x == 0) {
+        out[3] = v;
+    }
+}
+
+TEST(CUDABarrier, BarrierAnd) {
+    constexpr std::size_t n_bools = 4;
+    bool* out;
+    bool host_out[n_bools];
+
+    ASSERT_EQ(cudaMalloc(&out, n_bools * sizeof(bool)), cudaSuccess);
+
+    testBarrierAnd<<<1, 1024>>>(out);
+
+    ASSERT_EQ(cudaPeekAtLastError(), cudaSuccess);
+
+    ASSERT_EQ(cudaMemcpy(host_out, out, n_bools * sizeof(bool),
+                         cudaMemcpyDeviceToHost),
+              cudaSuccess);
+
+    ASSERT_FALSE(out[0]);
+    ASSERT_TRUE(out[1]);
+    ASSERT_FALSE(out[2]);
+    ASSERT_FALSE(out[3]);
+
+    ASSERT_EQ(cudaFree(out), cudaSuccess);
+}
+
+__global__ void testBarrierOr(bool* out) {
+    traccc::cuda::barrier bar;
+
+    bool v;
+
+    v = bar.blockOr(false);
+    if (threadIdx.x == 0) {
+        out[0] = v;
+    }
+
+    v = bar.blockOr(true);
+    if (threadIdx.x == 0) {
+        out[1] = v;
+    }
+
+    v = bar.blockOr(threadIdx.x % 2 == 0);
+    if (threadIdx.x == 0) {
+        out[2] = v;
+    }
+
+    v = bar.blockOr(threadIdx.x < 32);
+    if (threadIdx.x == 0) {
+        out[3] = v;
+    }
+}
+
+TEST(CUDABarrier, BarrierOr) {
+    constexpr std::size_t n_bools = 4;
+    bool* out;
+    bool host_out[n_bools];
+
+    ASSERT_EQ(cudaMalloc(&out, n_bools * sizeof(bool)), cudaSuccess);
+
+    testBarrierOr<<<1, 1024>>>(out);
+
+    ASSERT_EQ(cudaPeekAtLastError(), cudaSuccess);
+
+    ASSERT_EQ(cudaMemcpy(host_out, out, n_bools * sizeof(bool),
+                         cudaMemcpyDeviceToHost),
+              cudaSuccess);
+
+    ASSERT_FALSE(out[0]);
+    ASSERT_TRUE(out[1]);
+    ASSERT_TRUE(out[2]);
+    ASSERT_TRUE(out[3]);
+
+    ASSERT_EQ(cudaFree(out), cudaSuccess);
+}
+
+__global__ void testBarrierCount(int* out) {
+    traccc::cuda::barrier bar;
+
+    int v;
+
+    v = bar.blockOr(false);
+    if (threadIdx.x == 0) {
+        out[0] = v;
+    }
+
+    v = bar.blockOr(true);
+    if (threadIdx.x == 0) {
+        out[1] = v;
+    }
+
+    v = bar.blockOr(threadIdx.x % 2 == 0);
+    if (threadIdx.x == 0) {
+        out[2] = v;
+    }
+
+    v = bar.blockOr(threadIdx.x < 32);
+    if (threadIdx.x == 0) {
+        out[3] = v;
+    }
+}
+
+TEST(CUDABarrier, BarrierCount) {
+    constexpr std::size_t n_ints = 4;
+    int* out;
+    int host_out[n_ints];
+
+    ASSERT_EQ(cudaMalloc(&out, n_ints * sizeof(int)), cudaSuccess);
+
+    testBarrierCount<<<1, 1024>>>(out);
+
+    ASSERT_EQ(cudaPeekAtLastError(), cudaSuccess);
```

Comment:
On a synchronous kernel execution you might as well do:

```suggestion
    ASSERT_EQ(cudaDeviceSynchronize(), cudaSuccess);
```

Maybe in addition to your call. Though for "synchronous kernels" I've rather been using `cudaGetLastError()` instead. :thinking: For instance:

https://github.com/acts-project/vecmem/blob/main/tests/cuda/test_cuda_containers_kernels.cu#L59-L60